File: 프로그래머스/알고리즘 고득점 Kit/동적계획법(Dynamic Programming)/Level 3/정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> triangle) {
    vector<vector<int>> DP;
    int len = triangle.size();
    for(int i=0; i<len; i++){
        vector<int> v(len, 0);
        DP.push_back(v);
    }
    DP[0][0] = triangle[0][0];
    
    for(int i=1; i<len; i++){
        DP[i][0] = DP[i-1][0]+triangle[i][0];
        DP[i][i] = DP[i-1][i-1]+triangle[i][i];
        
        for(int j=1; j<i; j++){
            int a = DP[i-1][j]+triangle[i][j];
            int b = DP[i-1][j-1]+triangle[i][j];
            
            if(a >= b) DP[i][j] = a;
            else DP[i][j] = b;
        }
    }
    
    int max_num = DP[len-1][0];
    for(int i=1; i<len; i++)
        if(DP[len-1][i] > max_num)
            max_num = DP[len-1][i];
    
    return max_num;
}

File: 프로그래머스/알고리즘 고득점 Kit/동적계획법(Dynamic Programming)/Level 3/정수 삼각형/정수 삼각형.py
def solution(triangle):
    DP = [[0 for i in range(len(triangle))] for i in range(len(triangle))]
    DP[0][0] = triangle[0][0]

    for i in range(1, len(triangle)):
        DP[i][0] = DP[i-1][0]+triangle[i][0]
        DP[i][i] = DP[i-1][i-1]+triangle[i][i]

        for j in range(1, i):
            a = DP[i-1][j]+triangle[i][j]
            b = DP[i-1][j-1]+triangle[i][j]
            DP[i][j] = max(a, b)

    return max(DP[len(triangle)-1])
